{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import msprime\n",
    "import tskit\n",
    "import random\n",
    "import numpy as np\n",
    "import sklearn\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import classification_report\n",
    "import math\n",
    "\n",
    "# --- Demographic model ---\n",
    "\n",
    "def history_of_mexicans(len_seq, T, t_mig, N_e, n ):\n",
    "    demography = msprime.Demography()\n",
    "    \n",
    "    demography.add_population(name=\"AF\", initial_size=500000)\n",
    "    demography.add_population(name=\"EU\", initial_size=300000)\n",
    "    demography.add_population(name=\"NA\", initial_size=40000)\n",
    "    demography.add_population(name=\"MX\", initial_size=40000)\n",
    "\n",
    "    demography.add_population(name=\"OUT_AF\", initial_size=N_e)\n",
    "    demography.add_population(name=\"AF_ANC\", initial_size=N_e)\n",
    "\n",
    "    demography.add_population(name=\"NEAND\", initial_size=30000)\n",
    "    demography.add_population(name=\"ANCES\", initial_size=N_e)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    demography.add_admixture(time= T[3] , derived=\"MX\", ancestral=[\"AF\", \"EU\", \"NA\"], proportions=[0.05, 0.5, 0.45])\n",
    "    demography.add_population_split(time=T[2], derived=[\"EU\", \"NA\"], ancestral=\"OUT_AF\")\n",
    "    demography.add_mass_migration(time=t_mig, source='OUT_AF', dest='NEAND', proportion=0.05)\n",
    "    demography.add_population_split(time=T[1], derived=[\"AF\", \"OUT_AF\"], ancestral=\"AF_ANC\")\n",
    "    demography.add_population_split(time=T[0], derived=[\"AF_ANC\", \"NEAND\"], ancestral=\"ANCES\")\n",
    "\n",
    "    TS = msprime.sim_ancestry(\n",
    "        samples=\n",
    "        [       \n",
    "                msprime.SampleSet(1, ploidy=1, population='MX'),\n",
    "                msprime.SampleSet(n, ploidy=1, population='EU'), \n",
    "                msprime.SampleSet(n, ploidy=1, population='NA'),\n",
    "                msprime.SampleSet(n, ploidy=1, population='AF'),\n",
    "                msprime.SampleSet(25, ploidy=1, population='NEAND', time=2000)          \n",
    "               \n",
    "        ],\n",
    "    \n",
    "        ploidy=1,    \n",
    "        sequence_length=len_seq,\n",
    "        recombination_rate=2.5e-9, \n",
    "        demography=demography,\n",
    "        record_migrations=True, \n",
    "        random_seed=54321299,\n",
    "        num_replicates = 15\n",
    "                                \n",
    "    )\n",
    "    prop = []\n",
    "    for replicate_index, ts in enumerate(TS):\n",
    "        INDIVIDUAL = -1\n",
    "        for i in ts.tables.migrations:\n",
    "            if i.time == T[3] and i.dest == 0:\n",
    "                INDIVIDUAL = i.node\n",
    "                \n",
    "            if INDIVIDUAL != -1: \n",
    "                \n",
    "                proportions = np.array([0,  0, 0]) #proportions Af, Eu, NA\n",
    "                for i in ts.tables.migrations:\n",
    "                    if i.time == T[3] and i.node == INDIVIDUAL:\n",
    "                        if i.dest == 0:\n",
    "                            proportions[0] += i.right - i.left\n",
    "                        if i.dest == 1:\n",
    "                            proportions[1] += i.right - i.left\n",
    "                        if i.dest == 2:\n",
    "                            proportions[2] += i.right - i.left\n",
    "                proportions = proportions / sum(proportions)\n",
    "                break\n",
    "        if INDIVIDUAL != -1:\n",
    "            prop.append([proportions, ts])\n",
    "    \n",
    "    \n",
    "    s = 1\n",
    "    for i in range(len(prop)):\n",
    "        if prop[i][ 0][0] < 0.1 and prop[i][ 0][0] > 0.01 and prop[i][0][2] > 0.2 and prop[i][0][1] > 0.3:\n",
    "            s = prop[i][0][0]\n",
    "            ts_af = prop[i][1]\n",
    "            proportions = prop[i][0]\n",
    "            break\n",
    "    print('We generate mexican individual with Af-Eu-NA proportions = ', proportions)\n",
    "    \n",
    "    ts_af = msprime.sim_mutations(ts_af, rate=1.25e-8, random_seed=4321994)\n",
    "    \n",
    "    return ts_af, proportions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def clean_tracts(tractInit, CUT):\n",
    "    tract = np.copy(tractInit)\n",
    "    tract = tract/CUT\n",
    "    tract=tract.astype(int)\n",
    "    flag = True\n",
    "    while(flag):\n",
    "        flag=False\n",
    "        for i in range(len(tract)):\n",
    "            for j in range(len(tract)):\n",
    "                if not flag and tract[i,0]==tract[j,1]:\n",
    "                    tract[j,1]=tract[i,1]\n",
    "                    tract = np.delete(tract,i,0)\n",
    "                    flag=True\n",
    "    flag = True\n",
    "    while(flag):\n",
    "        flag=False\n",
    "        for i in range(len(tract)):\n",
    "            for j in range(i+1,len(tract)):\n",
    "                if tract[i,0]>tract[j,0]:\n",
    "                    save0=tract[i,0]\n",
    "                    save1=tract[i,1]\n",
    "                    tract[i,0]=tract[j,0]\n",
    "                    tract[i,1]=tract[j,1]\n",
    "                    tract[j,0]=save0\n",
    "                    tract[j,1]=save1\n",
    "                    flag=True\n",
    "    return tract\n",
    "\n",
    "      \n",
    "\n",
    "#несколько вспомогательных функций\n",
    "def connected(m):\n",
    "    for i in range(len(m)-1):\n",
    "        if m[i][1] == m[i+1][0]:\n",
    "            return True\n",
    "    return False\n",
    "        \n",
    "def remove_one(m):\n",
    "    mas = m\n",
    "    while connected(mas) == True:\n",
    "        for i in range(len(mas)-1):\n",
    "            if mas[i][1] == mas[i+1][0]:\n",
    "                mas[i][1] = mas[i+1][1]\n",
    "                mas.pop(i+1)\n",
    "                break\n",
    "    return mas\n",
    "\n",
    "\n",
    "#Вход: ts, название популяции, индивид(которого мы препарируем), время предка\n",
    "def get_migrating_tracts_ind(ts, pop, ind, T_anc):\n",
    "    mig = ts.tables.migrations\n",
    "    migration_int = []\n",
    "\n",
    "    for tree in ts.trees():  #перебираем все деревья. Как известно, каждому дереву отвечает участок днк  \n",
    "        anc_node = ind #выбираем мексиканца\n",
    "        while tree.time( tree.parent(anc_node) ) <= T_anc : #идем в прошлое до вершины anc_node по предкам нашего мексиканца, пока не наткнемся на миграцию неандертальцев\n",
    "            anc_node = tree.parent(anc_node)\n",
    "        migs = np.where(mig.node == anc_node)[0] #выбирем все строки, соответствующие заданному узлу\n",
    "\n",
    "        #идем по таблице миграций с anc_node и проверяем, чтобы миграции попадали в тот самый участок днк\n",
    "        for i in migs:\n",
    "\n",
    "            stroka = mig[i]\n",
    "            if stroka.time == T_anc and stroka.dest == pop and tree.interval.left >= stroka.left and tree.interval.right <= stroka.right:\n",
    "                migration_int.append([tree.interval.left, tree.interval.right])\n",
    "\n",
    "    migration_int2 = []\n",
    "    for i in range(len(migration_int)):\n",
    "        if migration_int[i][0] != migration_int[i][1]:\n",
    "            migration_int2.append(migration_int[i])\n",
    "    migration_int = migration_int2\n",
    "    \n",
    "    mi = remove_one(migration_int)\n",
    "    mi.sort()  \n",
    "\n",
    "    return mi\n",
    "\n",
    "\n",
    "\n",
    "#пересекаются ли интервалы. -1==нет\n",
    "def two_intersection(i1, i2):\n",
    "    if i1[0] > i2[1] or i1[1] < i2[0]:\n",
    "        return -1\n",
    "    if i1[0] >= i2[0] and i1[0] <= i2[1] and i1[1] >= i2[1]:\n",
    "        return [i1[0], i2[1]]\n",
    "    if i1[0] <= i2[0] and i1[1] <= i2[1] and i1[1] >= i2[0]:\n",
    "        return [i2[0],i1[1]]\n",
    "    if (i1[0] <= i2[0] and i1[1] >= i2[1]) or  (i2[0] <= i1[0] and i2[1] >= i1[1]):\n",
    "        return [max(i1[0],i2[0]), min(i1[1], i2[1])]\n",
    "    \n",
    "    \n",
    "#для двух наборов интервалов ищется их пересечение    \n",
    "def intersections(l1 , l2):\n",
    "    s = []\n",
    "    for i in l1:\n",
    "        for j in l2:\n",
    "            int_ij = two_intersection(i, j) \n",
    "            if int_ij != -1 and int_ij[0]!=int_ij[1]:\n",
    "                s.append(int_ij)\n",
    "\n",
    "    return s\n",
    "\n",
    "# лежит ли точка p в интервале i?\n",
    "def point_in_interval(p, i):\n",
    "    if p >= i[0] and p <= i[1]:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "# возвращает раскраску по состояниям в каждой позиции, тракты и [пропорции соврем днк в европейце, в американце ]\n",
    "\n",
    "def coloring(ts, ind,  t_mex):\n",
    "    \n",
    "    mig_neand=get_migrating_tracts_ind(ts, 6, ind, t_mig) # 6 номер популяции Неандертальца\n",
    "    \n",
    "    \n",
    "  \n",
    "    mig_eu = get_migrating_tracts_ind(ts,1, ind, t_mex) #европейцы\n",
    "    mig_na = get_migrating_tracts_ind(ts,2, ind, t_mex) # американцы(native americans)\n",
    "    mig_af = get_migrating_tracts_ind(ts,0, ind, t_mex) # Африканцы   \n",
    "        \n",
    "    \n",
    "    mig_eu.sort()\n",
    "    mig_na.sort()\n",
    "    mig_af.sort()    \n",
    "    mig_neand.sort()\n",
    "\n",
    "    \n",
    "    return mig_neand, mig_eu, mig_na, mig_af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# лежит ли точка p в интервале i?\n",
    "def point_in_interval(p, i):\n",
    "    if p >= i[0] and p <= i[1]:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "    \n",
    "#для двух наборов интервалов ищется их пересечение    \n",
    "def intersections(l1 , l2):\n",
    "    s = []\n",
    "    for i in l1:\n",
    "        for j in l2:\n",
    "            int_ij = two_intersection(i, j) \n",
    "            if int_ij != -1 and int_ij[0]!=int_ij[1]:\n",
    "                s.append(int_ij)\n",
    "\n",
    "    return s\n",
    "def coloring_by_snps(sites, mig_neand, mig_eu, mig_na, mig_af):\n",
    "    \n",
    "    \n",
    "    neand_eu = intersections(mig_neand , mig_eu) #неандерталец в европейце= пересечение неандертальских трактов и европейских\n",
    "  \n",
    "    neand_na = intersections(mig_neand, mig_na)\n",
    "    color = [[-1]*len(sites) for i in range(2)]\n",
    "    color[0] = sites\n",
    "    \n",
    "    \n",
    "    \n",
    "    s, s1,s2,s3,s4=[0,0,0,0,0] # число сайтов для каждого состояния \n",
    "    \n",
    "    for i in range(len(sites)):\n",
    "        for j in neand_eu:\n",
    "            if point_in_interval(color[0][i], j) == True:                \n",
    "                color[1][i] = 1            \n",
    "                \n",
    "    \n",
    "    \n",
    "    for i in range(len(sites)):        \n",
    "        for j in neand_na:            \n",
    "            if point_in_interval(color[0][i], j) == True:\n",
    "                color[1][i] = 3               \n",
    "                  \n",
    "            \n",
    "  \n",
    "    for i in range(len(sites)):\n",
    "        for j in mig_af:\n",
    "            if point_in_interval(color[0][i], j) == True:\n",
    "                color[1][i] = 4\n",
    "                                \n",
    "               \n",
    "     \n",
    "    for i in range(len(sites)):\n",
    "        for j in mig_eu:\n",
    "            if point_in_interval(color[0][i], j) == True and color[1][i] == -1:\n",
    "                color[1][i] = 0                              \n",
    "                \n",
    "              \n",
    "     \n",
    "    for i in range(len(sites)):\n",
    "        for j in mig_na:\n",
    "            if point_in_interval(color[0][i], j) == True and color[1][i] == -1:\n",
    "                color[1][i] = 2\n",
    "    return color\n",
    "            \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#массив из номеров позиций снипов\n",
    "def snps_positions(ts):\n",
    "\n",
    "    pos = []      \n",
    "    for v in ts.variants():\n",
    "        pos.append(v.site.position)\n",
    "\n",
    "    for i in range(len(pos)):\n",
    "        pos[i] = int(pos[i])\n",
    "        return pos\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def frequences(ts):\n",
    "    freq = []\n",
    "    print(type(freq))\n",
    "    for v in ts.variants():\n",
    "        freq.append([float(sum(v.genotypes[1:(n+1)]))/ n, float(sum(v.genotypes[(n+1):(2*n+1)]))/ n,\n",
    "                     float(sum(v.genotypes[(2*n+1) :(3*n+1)]))/ n, float(sum(v.genotypes[(3*n+1):(4*n+1)]))/ n]) \n",
    "    freq = np.array(freq)\n",
    "    freq = freq.transpose()\n",
    "    return freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def emission_probabilities(mu, frq, snps, t, t_africa):\n",
    "\n",
    "    sc_factor = math.exp(-mu * 2 *t)\n",
    "    sc_factor2 = math.exp(-mu * 2 * t_africa)\n",
    "    \n",
    "    b = []\n",
    "    theta = frq\n",
    "    print(len(snps))\n",
    "    \n",
    "    for tt in range(len(snps)):\n",
    "        \n",
    "        \n",
    "        b.append([[(1-theta[0, tt]) * sc_factor + theta[0, tt] * (1 - sc_factor),\n",
    "                   theta[0, tt] * sc_factor + (1-theta[0, tt]) * (1 - sc_factor)],\n",
    "                  [(1-theta[1, tt]) * sc_factor + theta[1, tt] * (1 - sc_factor),\n",
    "                   theta[1, tt] * sc_factor + (1-theta[1, tt]) * (1 - sc_factor)],              \n",
    "                  [(1-theta[2, tt]) * sc_factor2 + theta[2, tt] * (1 - sc_factor2),\n",
    "                   theta[2, tt] * sc_factor2 + (1-theta[2, tt]) * (1 - sc_factor2)]])\n",
    "\n",
    "        \n",
    "    return np.array(b)  \n",
    "        \n",
    "def observations(ts):\n",
    "        # наблюдаемый мексиканец \n",
    "    mexican = []\n",
    "    for v in ts.variants():\n",
    "        mexican.append(v.genotypes[0])\n",
    "    return np.array(mexican)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "N = 5 #number of hidden states\n",
    "K = 2 # 0 or 1 alleles\n",
    "\n",
    "#O - массив наблюдений из нулей и единичек; возвращаем наблюдение в позиции t\n",
    "def choose_state(o,t):\n",
    "    return o[t] \n",
    "\n",
    "# #этот кусок возвращает матрицу перехода, здесь d это как раз и есть расстояние между снипами,\n",
    "# r_m, r_a - недавняя и давняя рекомбинации\n",
    "# p_ee - вероятность перейти из европейца в европейца, p_ea - вероятность перейти из европейца в американца\n",
    "# (1 - p_ea - p_ee) - вероятность перейти из европейца в африканца\n",
    "#p_e_m - вероятность быть на современном куске, будучи в европейском куске\n",
    "#(1-p_e_m) - вероятность быть в древнем куске\n",
    "def transition_matrix(d, r_a, r_m, p_ee, p_e_m, p_a_m, p_ea, p_ae, p_aa, p_afe, p_afa):\n",
    "    tr_mat = np.zeros((5, 5))\n",
    "    \n",
    "    et_a = math.exp(-r_a * d)\n",
    "    et_m = math.exp(-r_m * d)\n",
    "    \n",
    "    tr_mat[0, 0] = et_a * et_m + ((1 - et_a) * et_m+(1 - et_m) * p_ee) * p_e_m\n",
    "    tr_mat[0, 1] = ((1 - et_a) * et_m + (1 - et_m) * p_ee) * (1 - p_e_m)\n",
    "    tr_mat[0, 2] = (1 - et_m) * p_ea * p_a_m\n",
    "    tr_mat[0, 3] = (1 - et_m) * p_ea * (1 - p_a_m)\n",
    "    tr_mat[0, 4] = (1 - et_m) * (1 - p_ea - p_ee)\n",
    "    \n",
    "    tr_mat[1, 0] = ((1 - et_a) * et_m + (1 - et_m) * p_ee) * p_e_m\n",
    "    tr_mat[1, 1] = et_a * et_m + ((1 - et_a) * et_m + (1 - et_m) * p_ee) * (1 - p_e_m)\n",
    "    tr_mat[1, 3] = (1 - et_m) * p_ea * (1 - p_a_m)\n",
    "    tr_mat[1, 2] = (1 - et_m) * p_ea * p_a_m\n",
    "    tr_mat[1, 4] = (1 - et_m) * (1 - p_ea -p_ee)   \n",
    "    \n",
    "    \n",
    "    tr_mat[2, 0] = (1 - et_m) * p_ae * p_e_m\n",
    "    tr_mat[2, 1] = (1 - et_m) * p_ae * (1 - p_e_m)\n",
    "    tr_mat[2, 2] = et_a * et_m + ((1-et_a) * et_m +(1 - et_m) * p_aa) * p_a_m\n",
    "    tr_mat[2, 3] = ((1 - et_a) * et_m + (1 - et_m) * p_aa) * (1 - p_a_m)\n",
    "    tr_mat[2, 4] = (1 - et_m) * (1 - p_ae - p_aa)\n",
    "    \n",
    "\n",
    "    tr_mat[3, 0] = (1 - et_m) * p_ae * p_e_m\n",
    "    tr_mat[3, 1] = (1 - et_m) * p_ae * (1 - p_e_m)\n",
    "    tr_mat[3, 3] = et_a * et_m + ((1-et_a) * et_m+(1 - et_m) * p_aa) * (1 - p_a_m)\n",
    "    tr_mat[3, 2] = ((1 - et_m) * et_m + (1 - et_m) * p_aa) * p_a_m\n",
    "    tr_mat[3, 4] = (1 - et_m) * (1 - p_ea - p_ee)\n",
    "    \n",
    "    tr_mat[4, 4] = et_m + (1-et_m) * (1 - p_afe - p_afa)\n",
    "    tr_mat[4, 0] = (1 - et_m) * p_afe * p_e_m\n",
    "    tr_mat[4, 1] = (1 - et_m) * p_afe * (1 - p_e_m)\n",
    "    tr_mat[4, 3] = (1 - et_m) * p_afa * (1 - p_a_m)\n",
    "    tr_mat[4, 2] = (1 - et_m) * p_afa * p_a_m\n",
    "    \n",
    "    return tr_mat\n",
    "\n",
    "# log-Viterbi algo for our matrices a-> many parameters\n",
    "def log_Viterbi(o, p,  b, pos):\n",
    "    \n",
    "    M = len(pos)\n",
    "    delta = np.zeros((N, M))\n",
    "    psi = [[0]*M for i in range(N)]\n",
    "    psi = np.array(psi)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    for i in range(N):\n",
    "        delta[i, 0] = math.log(b[0, i, choose_state(o, 0)]) + math.log(p[i])        \n",
    "    \n",
    "    for t in range(1, M):\n",
    "        \n",
    "        # define tr matrix\n",
    "\n",
    "        a = np.array(transition_matrix(pos[t]-pos[t-1], r_a, r_m, p_ee, p_e_m, p_a_m, p_ea, p_ae, p_aa, p_afe, p_afa))\n",
    "        \n",
    "        for j in range(N):\n",
    "            lst = []\n",
    "            for i in range(0, N):                \n",
    "                lst.append(delta[i, t-1] + math.log( a[i, j]))\n",
    "            delta[j, t] =  np.max (lst) + math.log( b[t,j, choose_state(o, t)])\n",
    "            psi[j, t] = np.argmax (lst )\n",
    "            \n",
    "    p = np.max(delta[:, M-1])\n",
    "    q = [0 for i in range(0,M)]\n",
    "    \n",
    "    q[M-1] = np.argmax(delta[:, M-1])    \n",
    "        \n",
    "    for t in range(M-2, 0, -1):\n",
    "        q[t] = psi[q[t+1], t+1]   \n",
    "    \n",
    "        \n",
    "    return q\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def emission_probabilities(mu, frq, snps, t, t_africa, t_neand_mig):\n",
    "\n",
    "    sc_factor = math.exp(-mu * 2 * t)\n",
    "    sc_factor2 = math.exp(-mu * 2 * t_africa)\n",
    "    sc_factor3 = math.exp(-mu * 2 * t_neand_mig)\n",
    "    \n",
    "    b = []\n",
    "    theta = frq\n",
    "    print(len(snps))\n",
    "    \n",
    "    for tt in range(len(snps)):\n",
    "        \n",
    "        \n",
    "        b.append([[(1-theta[0, tt]) * sc_factor + theta[0, tt] * (1 - sc_factor),\n",
    "                   theta[0, tt] * sc_factor + (1-theta[0, tt]) * (1 - sc_factor)],\n",
    "                  [(1-theta[3, tt]) * sc_factor3 + theta[3, tt] * (1 - sc_factor3),\n",
    "                   theta[3, tt] * sc_factor3 + (1-theta[3, tt]) * (1 - sc_factor3)],\n",
    "                  [(1-theta[1, tt]) * sc_factor + theta[1, tt] * (1 - sc_factor),\n",
    "                   theta[1, tt] * sc_factor + (1-theta[1, tt]) * (1 - sc_factor)],\n",
    "                  [(1-theta[3, tt]) * sc_factor3 + theta[3, tt] * (1 - sc_factor3),\n",
    "                   theta[3, tt] * sc_factor3 + (1-theta[3, tt]) * (1 - sc_factor3)],\n",
    "                  [(1-theta[2, tt]) * sc_factor + theta[2, tt] * (1 - sc_factor),\n",
    "                   theta[2, tt] * sc_factor + (1-theta[2, tt]) * (1 - sc_factor)]])\n",
    "\n",
    "        \n",
    "    return np.array(b)  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
