{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import msprime\n",
    "import tskit\n",
    "import random\n",
    "import numpy as np\n",
    "import sklearn\n",
    "import math\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import classification_report\n",
    "import random \n",
    "from random import randint, randrange\n",
    "from scipy.stats import poisson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Demographic model ---\n",
    "\n",
    "def history_of_mexicans(len_seq, T, t_mig, N_e, n ):\n",
    "    demography = msprime.Demography()\n",
    "    \n",
    "    demography.add_population(name=\"AF\", initial_size=500000)\n",
    "    demography.add_population(name=\"EU\", initial_size=300000)\n",
    "    demography.add_population(name=\"NA\", initial_size=40000)\n",
    "    demography.add_population(name=\"MX\", initial_size=40000)\n",
    "\n",
    "    demography.add_population(name=\"OUT_AF\", initial_size=N_e)\n",
    "    demography.add_population(name=\"AF_ANC\", initial_size=N_e)\n",
    "\n",
    "    demography.add_population(name=\"NEAND\", initial_size=30000)\n",
    "    demography.add_population(name=\"ANCES\", initial_size=N_e)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    demography.add_admixture(time= T[3] , derived=\"MX\", ancestral=[\"AF\", \"EU\", \"NA\"], proportions=[0.05, 0.5, 0.45])\n",
    "    demography.add_population_split(time=T[2], derived=[\"EU\", \"NA\"], ancestral=\"OUT_AF\")\n",
    "    demography.add_mass_migration(time=t_mig, source='OUT_AF', dest='NEAND', proportion=0.05)\n",
    "    demography.add_population_split(time=T[1], derived=[\"AF\", \"OUT_AF\"], ancestral=\"AF_ANC\")\n",
    "    demography.add_population_split(time=T[0], derived=[\"AF_ANC\", \"NEAND\"], ancestral=\"ANCES\")\n",
    "\n",
    "    TS = msprime.sim_ancestry(\n",
    "        samples=\n",
    "        [       \n",
    "                msprime.SampleSet(1, ploidy=1, population='MX'),\n",
    "                msprime.SampleSet(n, ploidy=1, population='EU'), \n",
    "                msprime.SampleSet(n, ploidy=1, population='NA'),\n",
    "                msprime.SampleSet(n, ploidy=1, population='AF'),\n",
    "                msprime.SampleSet(25, ploidy=1, population='NEAND', time=2000)          \n",
    "               \n",
    "        ],\n",
    "    \n",
    "        ploidy=1,    \n",
    "        sequence_length=len_seq,\n",
    "        recombination_rate=2.5e-9, \n",
    "        demography=demography,\n",
    "        record_migrations=True, \n",
    "        random_seed=5289,\n",
    "        num_replicates = 40\n",
    "                                \n",
    "    )\n",
    "    prop = []\n",
    "    for replicate_index, ts in enumerate(TS):\n",
    "        INDIVIDUAL = -1\n",
    "        for i in ts.tables.migrations:\n",
    "            if i.time == T[3] and i.dest == 0:\n",
    "                INDIVIDUAL = i.node\n",
    "                \n",
    "            if INDIVIDUAL != -1: \n",
    "                \n",
    "                proportions = np.array([0,  0, 0]) #proportions Af, Eu, NA\n",
    "                for i in ts.tables.migrations:\n",
    "                    if i.time == T[3] and i.node == INDIVIDUAL:\n",
    "                        if i.dest == 0:\n",
    "                            proportions[0] += i.right - i.left\n",
    "                        if i.dest == 1:\n",
    "                            proportions[1] += i.right - i.left\n",
    "                        if i.dest == 2:\n",
    "                            proportions[2] += i.right - i.left\n",
    "                proportions = proportions / sum(proportions)\n",
    "                break\n",
    "        if INDIVIDUAL != -1:\n",
    "            prop.append([proportions, ts])\n",
    "    \n",
    "    \n",
    "    s = 1\n",
    "    for i in range(len(prop)):\n",
    "        if prop[i][ 0][0] < 0.1 and prop[i][ 0][0] > 0.01 and prop[i][0][2] > 0.2 and prop[i][0][1] > 0.3:\n",
    "            s = prop[i][0][0]\n",
    "            ts_af = prop[i][1]\n",
    "            proportions = prop[i][0]\n",
    "            break\n",
    "    print('We generate mexican individual with Af-Eu-NA proportions = ', proportions)\n",
    "    \n",
    "    ts_af = msprime.sim_mutations(ts_af, rate=1.25e-8, random_seed=4321994, discrete_genome=False)\n",
    "    \n",
    "    return ts_af, proportions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def clean_tracts(tractInit, CUT):\n",
    "    tract = np.copy(tractInit)\n",
    "    tract = tract/CUT\n",
    "    tract=tract.astype(int)\n",
    "    flag = True\n",
    "    while(flag):\n",
    "        flag=False\n",
    "        for i in range(len(tract)):\n",
    "            for j in range(len(tract)):\n",
    "                if not flag and tract[i,0]==tract[j,1]:\n",
    "                    tract[j,1]=tract[i,1]\n",
    "                    tract = np.delete(tract,i,0)\n",
    "                    flag=True\n",
    "    flag = True\n",
    "    while(flag):\n",
    "        flag=False\n",
    "        for i in range(len(tract)):\n",
    "            for j in range(i+1,len(tract)):\n",
    "                if tract[i,0]>tract[j,0]:\n",
    "                    save0=tract[i,0]\n",
    "                    save1=tract[i,1]\n",
    "                    tract[i,0]=tract[j,0]\n",
    "                    tract[i,1]=tract[j,1]\n",
    "                    tract[j,0]=save0\n",
    "                    tract[j,1]=save1\n",
    "                    flag=True\n",
    "    return tract\n",
    "\n",
    "      \n",
    "\n",
    "#несколько вспомогательных функций\n",
    "def connected(m):\n",
    "    for i in range(len(m)-1):\n",
    "        if m[i][1] == m[i+1][0]:\n",
    "            return True\n",
    "    return False\n",
    "        \n",
    "def remove_one(m):\n",
    "    mas = m\n",
    "    while connected(mas) == True:\n",
    "        for i in range(len(mas)-1):\n",
    "            if mas[i][1] == mas[i+1][0]:\n",
    "                mas[i][1] = mas[i+1][1]\n",
    "                mas.pop(i+1)\n",
    "                break\n",
    "    return mas\n",
    "\n",
    "\n",
    "#Вход: ts, название популяции, индивид(которого мы препарируем), время предка\n",
    "def get_migrating_tracts_ind(ts, pop, ind, T_anc):\n",
    "    mig = ts.tables.migrations\n",
    "    migration_int = []\n",
    "\n",
    "    for tree in ts.trees():  #перебираем все деревья. Как известно, каждому дереву отвечает участок днк  \n",
    "        anc_node = ind #выбираем мексиканца\n",
    "        while tree.time( tree.parent(anc_node) ) <= T_anc : #идем в прошлое до вершины anc_node по предкам нашего мексиканца, пока не наткнемся на миграцию неандертальцев\n",
    "            anc_node = tree.parent(anc_node)\n",
    "        migs = np.where(mig.node == anc_node)[0] #выбирем все строки, соответствующие заданному узлу\n",
    "\n",
    "        #идем по таблице миграций с anc_node и проверяем, чтобы миграции попадали в тот самый участок днк\n",
    "        for i in migs:\n",
    "\n",
    "            stroka = mig[i]\n",
    "            if stroka.time == T_anc and stroka.dest == pop and tree.interval.left >= stroka.left and tree.interval.right <= stroka.right:\n",
    "                migration_int.append([tree.interval.left, tree.interval.right])\n",
    "\n",
    "    migration_int2 = []\n",
    "    for i in range(len(migration_int)):\n",
    "        if migration_int[i][0] != migration_int[i][1]:\n",
    "            migration_int2.append(migration_int[i])\n",
    "    migration_int = migration_int2\n",
    "    \n",
    "    mi = remove_one(migration_int)\n",
    "    mi.sort()  \n",
    "\n",
    "    return mi\n",
    "\n",
    "\n",
    "\n",
    "#пересекаются ли интервалы. -1==нет\n",
    "def two_intersection(i1, i2):\n",
    "    if i1[0] > i2[1] or i1[1] < i2[0]:\n",
    "        return -1\n",
    "    if i1[0] >= i2[0] and i1[0] <= i2[1] and i1[1] >= i2[1]:\n",
    "        return [i1[0], i2[1]]\n",
    "    if i1[0] <= i2[0] and i1[1] <= i2[1] and i1[1] >= i2[0]:\n",
    "        return [i2[0],i1[1]]\n",
    "    if (i1[0] <= i2[0] and i1[1] >= i2[1]) or  (i2[0] <= i1[0] and i2[1] >= i1[1]):\n",
    "        return [max(i1[0],i2[0]), min(i1[1], i2[1])]\n",
    "    \n",
    "    \n",
    "#для двух наборов интервалов ищется их пересечение    \n",
    "def intersections(l1 , l2):\n",
    "    s = []\n",
    "    for i in l1:\n",
    "        for j in l2:\n",
    "            int_ij = two_intersection(i, j) \n",
    "            if int_ij != -1 and int_ij[0]!=int_ij[1]:\n",
    "                s.append(int_ij)\n",
    "\n",
    "    return s\n",
    "\n",
    "# лежит ли точка p в интервале i?\n",
    "def point_in_interval(p, i):\n",
    "    if p >= i[0] and p <= i[1]:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "# возвращает раскраску по состояниям в каждой позиции, тракты и [пропорции соврем днк в европейце, в американце ]\n",
    "\n",
    "def coloring(ts, ind,  t_mex):\n",
    "    \n",
    "    mig_neand=get_migrating_tracts_ind(ts, 6, ind, t_mig) # 6 номер популяции Неандертальца\n",
    "    \n",
    "    \n",
    "  \n",
    "    mig_eu = get_migrating_tracts_ind(ts,1, ind, t_mex) #европейцы\n",
    "    mig_na = get_migrating_tracts_ind(ts,2, ind, t_mex) # американцы(native americans)\n",
    "    mig_af = get_migrating_tracts_ind(ts,0, ind, t_mex) # Африканцы   \n",
    "        \n",
    "    \n",
    "    mig_eu.sort()\n",
    "    mig_na.sort()\n",
    "    mig_af.sort()    \n",
    "    mig_neand.sort()\n",
    "\n",
    "    \n",
    "    return mig_neand, mig_eu, mig_na, mig_af\n",
    "\n",
    "# функция, переводящая состояния-гаплотипы в состояния европа-америка-африка\n",
    "def haplo_states_to_eu_am_af_states(m):\n",
    "    \n",
    "    m2 = []\n",
    "    \n",
    "    st0 = [i for i in range(n_haplo)]\n",
    "    st1 = [i for i in range(n_haplo, 2* n_haplo)]\n",
    "    st2 = [i for i in range(2*n_haplo, 3*n_haplo)]\n",
    "    st3 = [i for i in range(3*n_haplo, 3*n_haplo+n_neand)]\n",
    "    \n",
    "    for i in range(len(m)):\n",
    "        if m[i] in st0:\n",
    "            m2.append(0)\n",
    "        if  m[i] in st1:\n",
    "             m2.append(1)\n",
    "        if  m[i] in st2:\n",
    "             m2.append(2)  \n",
    "        if  m[i] in st3:\n",
    "             m2.append(3) \n",
    "    \n",
    "    return m2\n",
    "\n",
    "def coloring_by_snps_haplo(sites, mig_neand, mig_eu, mig_na, mig_af):\n",
    "    \n",
    "    \n",
    "\n",
    "    color = [[-1]*len(sites) for i in range(2)]\n",
    "    color[0] = sites\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "    \n",
    "    for i in range(len(sites)):\n",
    "        for j in mig_neand:\n",
    "            if point_in_interval(color[0][i], j) == True:                \n",
    "                color[1][i] = 3            \n",
    "                \n",
    "\n",
    "  \n",
    "    for i in range(len(sites)):\n",
    "        for j in mig_af:\n",
    "            if point_in_interval(color[0][i], j) == True:\n",
    "                color[1][i] = 2\n",
    "                                \n",
    "               \n",
    "     \n",
    "    for i in range(len(sites)):\n",
    "        for j in mig_eu:\n",
    "            if point_in_interval(color[0][i], j) == True and color[1][i] == -1:\n",
    "                color[1][i] = 0                              \n",
    "                \n",
    "              \n",
    "     \n",
    "    for i in range(len(sites)):\n",
    "        for j in mig_na:\n",
    "            if point_in_interval(color[0][i], j) == True and color[1][i] == -1:\n",
    "                color[1][i] = 1\n",
    "    return color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# выбираем гаплотипы для состояний. важно, чтобы они не повторялись\n",
    "\n",
    "#если в массиве есть повторяющиеся значения, то возвращает True, иначе False\n",
    "def repetitions(arr):\n",
    "\n",
    "    for elem in arr:\n",
    "        if arr.count(elem) > 1:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "\n",
    "# массив m - массив массивов с номерами(почему не просто массив? так вышло.) Возвращает массив из значений True/False, \n",
    "#отвечающих за повторяющиеся значения. \n",
    "def repetitions2(m):\n",
    "    return [repetitions(m[0]), repetitions(m[1]), repetitions(m[2])]\n",
    "\n",
    "#возвращает массив номеров гаплотипов для ХММ \n",
    "# [randrange(0, n) for i in range(n_haplo)] выбирает n_haplo чисел из интервала [0, n-1]\n",
    "def haplotypes_random_numbers(n_haplo, n,n_neand):\n",
    "    m = [[randrange(1, n+1) for i in range(n_haplo)], [randrange(n+1, 2*n+1) for i in range(n_haplo)], \n",
    "         [randrange(2*n+1, 3*n+1) for i in range(n_haplo)], [randrange(3*n+1, 3*n+1+n_neand) for i in range(n_neand)]]\n",
    "    \n",
    "    # пока есть хоть одно повторяющееся значение, делаем...\n",
    "    while True in repetitions2(m):\n",
    "        m = [[randrange(1, n+1) for i in range(n_haplo)], [randrange(n+1, 2*n+1) for i in range(n_haplo)], \n",
    "             [randrange(2*n+1, 3*n+1) for i in range(n_haplo)], [randrange(3*n+1, 3*n+1+n_neand) for i in range(n_neand)]]\n",
    "       \n",
    "    m = [sorted(m[0]), sorted(m[1]), sorted(m[2]), sorted(m[3])]\n",
    "    print('Индексы индивидуумов', m)\n",
    "    mm = []\n",
    "    for i in range(len(m)):\n",
    "        for j in range(len(m[i])):\n",
    "            mm.append(m[i][j])\n",
    "    return mm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#массив из номеров позиций снипов\n",
    "def snps_positions(ts):\n",
    "\n",
    "    pos = []      \n",
    "    for v in ts.variants():\n",
    "        pos.append(v.site.position)\n",
    "\n",
    "\n",
    "    return np.array(pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def observations_haplo_window(ts, L,   hap_nums):\n",
    "    \n",
    "    \n",
    "    n_h = len(hap_nums)\n",
    "    f = L\n",
    "    d_in_snps = np.zeros(n_h, dtype = np.int8)\n",
    "    d_in_one_wind = []\n",
    "    d_in_wind = []\n",
    "    \n",
    "\n",
    "    \n",
    "    for v in ts.variants():\n",
    "        \n",
    "        if v.site.position < f:\n",
    "            d_in_snps = [abs(v.genotypes[0]-v.genotypes[i]) for i in hap_nums]\n",
    "            \n",
    "            d_in_one_wind.append(d_in_snps)\n",
    "            \n",
    "        else:\n",
    "            f += L\n",
    "            \n",
    "            d_in_one_wind = np.array(d_in_one_wind)\n",
    "            d_in_wind.append( np.sum(d_in_one_wind, axis=0))\n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "            d_in_one_wind = []\n",
    "            d_in_snps = [abs(v.genotypes[0]-v.genotypes[i]) for i in hap_nums]\n",
    "            d_in_one_wind.append(d_in_snps)\n",
    "            \n",
    "    \n",
    "    d_in_one_wind = np.array(d_in_one_wind)\n",
    "    d_in_wind.append( np.sum(d_in_one_wind, axis=0))\n",
    "    return np.array(d_in_wind)                \n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def emission_haplo_window(o, L, t_mx,  t_nd, mu, n_haplo, n_nd):\n",
    "    # определим количество разных наблюдений\n",
    "    o = np.array(o)    \n",
    "    n_o = np.max(o)\n",
    "    b=[]\n",
    "    \n",
    "    lambda_nd = mu * L * t_nd\n",
    "    lambda_mx = mu * L * t_mx\n",
    "    \n",
    "    b_nd = [poisson.pmf(k=i, mu=lambda_nd) for i in range(n_o+1)]\n",
    "    b_mx = [poisson.pmf(k=i, mu=lambda_mx) for i in range(n_o+1)]\n",
    "          \n",
    "    \n",
    "    for i in range(3*n_haplo+n_nd):\n",
    "        if i < 3*n_haplo:\n",
    "            b.append(b_mx)\n",
    "\n",
    "        else:\n",
    "            \n",
    "            b.append(b_nd)\n",
    "    \n",
    "    \n",
    "    return np.array(b)  \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def choose_state(o,t,i):\n",
    "    return o[t][i] \n",
    "\n",
    "def log_Viterbi(o, p,  b, a):\n",
    "    \n",
    "    M = len(o)\n",
    "    N = N_haplo*3+n_neand\n",
    "    delta = np.zeros((N_haplo*3+n_neand, M))\n",
    "    psi = [[0]*M for i in range(N)]\n",
    "    psi = np.array(psi)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    for i in range(N):\n",
    "        delta[i, 0] = math.log(b[ i, choose_state(o, 0, i)]) + math.log(p[i])        \n",
    "    \n",
    "    for t in range(1, M):\n",
    "        \n",
    "        # define tr matrix\n",
    "\n",
    "        \n",
    "        \n",
    "        for j in range(N):\n",
    "            lst = []\n",
    "            for i in range(0, N):                \n",
    "                lst.append(delta[i, t-1] + math.log( a[i, j]))\n",
    "            delta[j, t] =  np.max (lst) + math.log( b[j, choose_state(o, t, j)])\n",
    "            psi[j, t] = np.argmax (lst )\n",
    "            \n",
    "    p = np.max(delta[:, M-1])\n",
    "    q = [0 for i in range(0,M)]\n",
    "    \n",
    "    q[M-1] = np.argmax(delta[:, M-1])    \n",
    "        \n",
    "    for t in range(M-2, 0, -1):\n",
    "        q[t] = psi[q[t+1], t+1]   \n",
    "    \n",
    "        \n",
    "    return q\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tr_mat_4x4(r,L,t_nd, t_mx, p_eu, p_na, p_nd):\n",
    "    a=np.zeros((4,4),dtype='f')\n",
    "    \n",
    "    a[0][1] = t_mx * r * L * p_na * (1-p_nd)\n",
    "    a[0][2] = t_mx * r * L * (1-p_eu-p_na)\n",
    "    a[0][3] = t_nd * r * L * (p_eu+p_na)* p_nd\n",
    "    a[0][0] = 1- a[0][1]-a[0][2]-a[0][3]\n",
    "    \n",
    "    a[1][0] = t_mx * r * L * p_eu * (1-p_nd)    \n",
    "    a[1][2] = t_mx * r * L * (1-p_eu-p_na)\n",
    "    a[1][3] = t_nd * r * L *  (p_eu+p_na)*p_nd\n",
    "    a[1][1] = 1- a[1][0]-a[1][2]-a[1][3]\n",
    "    \n",
    "    a[2][0] = t_mx * r * L * p_eu * (1-p_nd)\n",
    "    a[2][1] = t_mx * r * L * p_na * (1-p_nd)\n",
    "    a[2][3] = t_nd * r * L * (p_eu+p_na)* p_nd\n",
    "    a[2][2] = 1- a[2][0]-a[2][1]-a[2][3]    \n",
    "\n",
    "    a[3][0] = t_nd * r * L * p_eu * (1-p_nd)\n",
    "    a[3][1] = t_nd * r * L * p_na * (1-p_nd)\n",
    "    a[3][2] = t_mx * r * L * (1-p_eu-p_na)\n",
    "    a[3][3] = 1- a[3][0]-a[3][1]-a[3][2]   \n",
    "    \n",
    "    return a\n",
    "\n",
    "def transition_matrix_haplo(r, L, N_haplo, n_neand, t_nd, t_mx, p_eu, p_na, p_nd):\n",
    "    AA = tr_mat_4x4(r,L,t_nd, t_mx, p_eu, p_na, p_nd)\n",
    "    \n",
    "    tr_mat_haplo = np.zeros((3 * N_haplo+n_neand, 3 * N_haplo+n_neand))\n",
    "    \n",
    "    for i in range(3):\n",
    "        for j in range(3):\n",
    "            for k in range(N_haplo):\n",
    "                for l in range(N_haplo):\n",
    "                    tr_mat_haplo [N_haplo * i + k][N_haplo * j+l] = AA[i][j]/ N_haplo\n",
    "                    \n",
    "    for k in range(n_neand):\n",
    "        for l in range(n_neand):\n",
    "            tr_mat_haplo [3*N_haplo  + k][3*N_haplo +l] = AA[3][3]/ n_neand\n",
    "            \n",
    "            \n",
    "    for j in range(3):\n",
    "        for k in range(n_neand):\n",
    "            for l in range(N_haplo):\n",
    "                tr_mat_haplo [3*N_haplo  + k][j*N_haplo +l] = AA[3][j]/ n_neand\n",
    "                \n",
    "                \n",
    "    for i in range(3):\n",
    "        for k in range(N_haplo):\n",
    "            for l in range(n_neand):\n",
    "            \n",
    "                tr_mat_haplo [i*N_haplo  + k][ 3*N_haplo+l] = AA[i][3]/ n_neand\n",
    " \n",
    "    return np.array(tr_mat_haplo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def initial(n_haplo, n_neand):\n",
    "    p = np.zeros(3*n_haplo+n_neand, dtype = 'f')\n",
    "    for i in range(3*n_haplo):\n",
    "        p[i] = 0.25/(n_haplo)\n",
    "    for i in range(3*n_haplo, 3*n_haplo+n_neand):\n",
    "        p[i] = 0.25/(n_neand)\n",
    "        \n",
    "    return p\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def vit_st_window_to_snp_st(snps, L, vit_states_window):\n",
    "    m=[]\n",
    "    k=0\n",
    "    for i in range(len(snps)):\n",
    "        \n",
    "        \n",
    "        if snps[i] < k*L +L:\n",
    "            m.append(vit_states_window[k])\n",
    "        else:\n",
    "            k+=1\n",
    "            m.append(vit_states_window[k])\n",
    "    return m\n",
    "\n",
    "print(vit_st_window_to_snp_st([5,6,15,19,21,24,35,36], 10, [2,3,2,5]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#demographic hystory and ancestry of tracts\n",
    "\n",
    "time_units = 1000 / 25 #number of generations per 1000 years   \n",
    "T = [600 * time_units, 70 * time_units, 30 * time_units,  time_units] \n",
    "t_mig = 50 * time_units \n",
    "\n",
    "r = 2.5e-9\n",
    "mu=1.25e-8\n",
    "N_e = 50000 \n",
    "n = 250 \n",
    "ind_number = 0\n",
    "len_sequence = 12e7\n",
    "t_mex = T[3]\n",
    "\n",
    "ts, proportions = history_of_mexicans(len_sequence, T, t_mig, N_e, n)\n",
    "\n",
    "tractsND = get_migrating_tracts_ind(ts, 6, ind_number, t_mig)\n",
    "tractsEU = get_migrating_tracts_ind(ts,1, ind_number, t_mex)\n",
    "tractsAS = get_migrating_tracts_ind(ts,2, ind_number, t_mex)\n",
    "tractsAF =get_migrating_tracts_ind(ts,0, ind_number, t_mex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_haplo=...\n",
    "N_haplo=...\n",
    "n_neand=..\n",
    "L = ...\n",
    "haplotypes_numbers = haplotypes_random_numbers(n_haplo, n, n_neand)\n",
    "O = observations_haplo_window(ts, L,   haplotypes_numbers)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#detecting only neanderthal states\n",
    "\n",
    "b = emission_haplo_window(O, L, t_mex,  t_mig, mu, n_haplo, n_neand)\n",
    "p=initial(n_haplo,n_neand)\n",
    "A = transition_matrix_haplo(r,L,N_haplo, n_neand, t_mig, t_mex, 0.5, 0.45, 0.05)  \n",
    "vit_states_window = log_Viterbi(O, p,  b, A)\n",
    "states_window = haplo_states_to_eu_am_af_states(vit_states_window)\n",
    "state_vits_snp = vit_st_window_to_snp_st(snps, L, states_window)\n",
    "\n",
    "\n",
    "real_states = coloring_by_snps_haplo(snps, tractsND, tractsEU, tractsAS, tractsAF)[1]\n",
    "print(sklearn.metrics.classification_report(real_states,state_vits_snp,  digits=5))\n",
    "print(sklearn.metrics.confusion_matrix(real_states,state_vits_snp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
